llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -12786,6 +12788,193 @@ struct AAAddressSpaceCallSiteArgument final : AAAddressSpaceImpl {
 };
 } // namespace
 
+/// ------------------------ No Alias Address Space  ---------------------------
+// This attribute assumes flat address space can alias all other address space
```

Comment:
Honestly, I think this and AAAddressSpace's impl class can be merged such that one impl class can serve two AAs. Can you try and see that is feasible? I'd say 80% of the code is similar.